// pui-react-lists/lists.js
/*(c) Copyright 2015 Pivotal Software, Inc. All Rights Reserved.*/
'use strict';

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var React = require('react');

var types = React.PropTypes;
require('pui-css-lists');

var ListItem = function (_React$Component) {
  (0, _inherits3.default)(ListItem, _React$Component);

  function ListItem() {
    (0, _classCallCheck3.default)(this, ListItem);
    return (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(ListItem).apply(this, arguments));
  }

  (0, _createClass3.default)(ListItem, [{
    key: 'render',
    value: function render() {
      return React.createElement('li', this.props);
    }
  }]);
  return ListItem;
}(React.Component);

function defList(tagName, spacingType, classNames, childClassNames) {
  var _class, _temp;

  return _temp = _class = function (_React$Component2) {
    (0, _inherits3.default)(_class, _React$Component2);

    function _class() {
      (0, _classCallCheck3.default)(this, _class);
      return (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(_class).apply(this, arguments));
    }

    (0, _createClass3.default)(_class, [{
      key: 'render',
      value: function render() {
        var _props = this.props;
        var className = _props.className;
        var spacing = _props.spacing;
        var children = _props.children;
        var checked = _props.checked;
        var unstyled = _props.unstyled;
        var divider = _props.divider;
        var others = (0, _objectWithoutProperties3.default)(_props, ['className', 'spacing', 'children', 'checked', 'unstyled', 'divider']);

        var classes = (0, _classnames2.default)(classNames(this.props), className, spacing && '' + spacingType + spacing);
        if (childClassNames) {
          children = React.Children.map(children, function (child) {
            return React.cloneElement(child, { className: childClassNames });
          });
        }
        return tagName === 'ul' ? React.createElement(
          'ul',
          (0, _extends3.default)({ className: classes }, others),
          children
        ) : tagName === 'ol' ? React.createElement(
          'ol',
          (0, _extends3.default)({ className: classes }, others),
          children
        ) : null;
      }
    }]);
    return _class;
  }(React.Component), _class.propTypes = {
    spacing: types.oneOf(['n', 's', 'm', 'l', 'xl']),
    className: types.string,
    checked: types.bool,
    unstyled: types.bool,
    divider: types.bool
  }, _temp;
}

var UnorderedList = defList('ul', 'lv', function (_ref) {
  var unstyled = _ref.unstyled;
  var checked = _ref.checked;
  return (0, _classnames2.default)({
    'list-unordered': !unstyled && !checked,
    'list-unstyled': unstyled,
    'list-checked': checked
  });
});

var OrderedList = defList('ol', 'lv', function (_ref2) {
  var unstyled = _ref2.unstyled;
  return (0, _classnames2.default)({ 'list-unstyled': unstyled });
});

var InlineList = defList('ul', 'lh', function (_ref3) {
  var divider = _ref3.divider;
  return (0, _classnames2.default)('list-inline', { 'list-inline-divider': divider });
});

var GroupList = defList('ul', 'lv', function () {
  return (0, _classnames2.default)('list-group');
}, 'list-group-item');

var GroupListInverse = defList('ul', 'lv', function () {
  return (0, _classnames2.default)('list-group-inverse');
}, 'list-group-item');

var StepList = defList('ol', 'lh', function () {
  return (0, _classnames2.default)('list-steps');
});

var BreadcrumbList = defList('ul', 'lh', function () {
  return (0, _classnames2.default)('list-breadcrumb');
});

module.exports = {
  ListItem: ListItem,
  UnorderedList: UnorderedList,
  OrderedList: OrderedList,
  InlineList: InlineList,
  GroupList: GroupList,
  GroupListInverse: GroupListInverse,
  StepList: StepList,
  BreadcrumbList: BreadcrumbList
};
